guava/src/com/google/common/io/BaseEncoding.java

```diff
@@ -468,19 +477,21 @@ public static BaseEncoding base16() {
       this.validPadding = validPadding;
     }
 
-    char encode(int bits) {
+    char encode(@IndexFor("this.chars") int bits) {
       return chars[bits];
     }
 
+    @SuppressWarnings("array.access.unsafe") // The index used is smaller than charsPerChunk, which is the length of validPadding
     boolean isValidPaddingStartPosition(int index) {
       return validPadding[index % charsPerChunk];
     }
 
+    @SuppressWarnings("array.access.unsafe") // decodabet won't be accessed with an invalid index because it was previously checked.
```

Comment:
This seems to imply that the argument `ch` should be `@IndexFor`, and you should suppress warnings at the callsites if necessary.

Reply:
A `char` is always non-negative and it was previously checked that it is less than `Ascii.MAX + 1`, which is the length of `decodabet`. The warning was just about the index possibly being < 0, which is impossible because the index is a `char`.